dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,153 @@
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(count-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+def get_args():
+    """Creates and returns a kafka consumer"""
+    parser = ArgumentParser()
+    parser.add_argument("--topic", default="default",
+                        help="The name of the kafka topic from which to read")
+    parser.add_argument("--group", default="my-group",
+                        help="The kafka group in which the topic exists")
+    parser.add_argument("--server", default="localhost:9092",
+                        help="Name and port of the kafka server (e.g. localhost:9092)")
+    arguments = parser.parse_args()
+
+    # To consume latest messages and auto-commit offsets
+    return arguments.topic, arguments.group, arguments.server
+
+def redraw(figures):
+    """Redraw all figures from a dict of figures"""
+    for key in figures:
+        figures[key].canvas.draw()
+        figures[key].canvas.flush_events()
+
+
+def update_data(data, value):
+    """Update the list of data, given some json input"""
+    try:
+        update = loads(value)
+    except decoder.JSONDecodeError as problem:
+        print("Error: Flawed JSON in value: ",
+              value)
+        print("Error message: ", problem)
+        update = ""
+    data.update(update)
+
+
+def graph(topic, group, server):
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = initscr()
+    cbreak()
+    screen.nodelay(True)
+    screen.getch()
+
+    print("Generic Graphing Consumer")
+    print("Press any key to exit...")
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    count = 0
+    pyplot.ion() #Interactive mode on
+    configuration = {"bootstrap.servers": server,
+                     "group.id": group,
+                     "default.topic.config" : {
+                         "auto.offset.reset": "earliest"},
+                     "enable.auto.commit": False}
+    consumer = Consumer(configuration)
+    consumer.subscribe([topic])
+    keyboard_input = ERR
+
+    while True:
+        if keyboard_input != ERR:
+            break
+        keyboard_input = screen.getch()
+        message = consumer.consume(num_messages=1, timeout=0.2)
+        if (message is None) or (message == []):
+            redraw(figures)
+        elif message[0].error():
+            print("Something went wrong with the read!")
+            print(message.error())
+        else:
+            update_data(data, message[0].value())
+            count += 1
+
+            populate(line, data, count)
+
+            for key in line:
+                while len(line[key]) < count:
+                    line[key].append(0) #Variable went missing from data.
```

Comment:
This catches the specific case where a variable with subvariables (example: {"A": {"B": 3, "C": 5}}) abruptly turns into a variable without subvariables (example: {"A": 3}) in the input.

I'll think about other ways to handle it, but it will need to be specifically handled.

Reply:
What would cause that abrupt change?

Reply:
Nothing that I can think of. However, this is supposed to take anything that's valid JSON and graph it; so it has to handle even input that I can't see a reason for.

Reply:
Have considered the case for moving this check elsewhere. It does need to be after the call to the populate function (which is right before this line), and I don't think it can be handled cleanly inside the populate function  (largely because the populate function is recursive; trying to handle it in there will require an if-statement to ensure that we're in the top level of the recursion and then an additional loop through the keys of the variable dict). So I don't really see anywhere to put this that's cleaner than where it is.